### DISCUSSION TITLE
Can I have Github pull-pr bottle the big_sur version for arm64_big_sur, since its build dependencies are not available for arm64?

### DISCUSSION THREAD

**Discussion by aryairani** [Homebrew/discussions#1641] [Tap maintenance and Homebrew development] [answered]
Or configure an `arm64_big_sur` install to pour `big_sur` bottles?

I'm just looking to make installation easy for users.

(I even wouldn't mind reusing the `catalina:` bottle, which gets created by the standard tap-new scripts, for `big_sur:`.)

Are the scripts created with `tap-new` maybe out of date? the ones at https://github.com/Homebrew/homebrew-core/blob/master/.github/workflows/tests.yml seem more sophisticated, `CHANGE_ARCH_TO_ARM` for example, but I'm not sure if they are drop-in replacements for custom taps.

**Comment by SMillerDev** [comment, marked as answer]
You can't use x86 software natively on an ARM mac. But if the user uses brew with rosetta it'll use the normal x86 binaries.

**Reply by aryairani** [reply]
Hi @SMillerDev — thanks for the reply.  I didn't follow though, and for what it's worth I'm on an ARM mac and trying to get some bottles deployed.  Do you mean, "if the user is using the rosetta install of brew, e.g. in `/usr/local/`, then it will look for `big_sur` bottles instead of `arm64_big_sur` bottles."?

I noticed that the github actions that get set up by `tap-new` don't build for `big_sur` at all, only `catalina` and `linux_x64`; whereas hombrew-core seems to build for all platforms.  Can you speak to that?  Or I can open a separate question if that's better.

Thanks!

**Reply by nandahkrishna** [reply]
> "if the user is using the rosetta install of brew, e.g. in `/usr/local/`, then it will look for `big_sur` bottles instead of `arm64_big_sur` bottles."?

Yes.

> I noticed that the github actions that get set up by `tap-new` don't build for `big_sur` at all, only `catalina` and `linux_x64`; whereas hombrew-core seems to build for all platforms. Can you speak to that? Or I can open a separate question if that's better.

The `tap-new` command sets up an action using a template, and that template uses `macos-latest` and `ubuntu-latest`. On GitHub Actions, `macos-latest` is Catalina (see https://github.com/actions/virtual-environments#available-environments). You can add `macos-11` to the workflow to build on Big Sur on GitHub Actions.

As for ARM bottles, I don't think GitHub Actions has M1 runners yet. You could bottle your formulae locally on your M1 Mac and upload the bottles instead.

**Reply by aryairani** [reply]
The Homebrew org seems to have its own `11-arm` runners; how did you do that?  Is that something I can use for automation, even though I am trying to publish formula for our own software (frowned upon, I guess?)?

**Reply by nandahkrishna** [reply]
We use [self-hosted runners](https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners). It isn't possible to use Homebrew's CI to publish bottles in your personal tap. If your formula is for popular open-source software (for GitHub repositories that would mean 75+ stars or 30+ forks/watchers), you can submit a PR.

> frowned upon, I guess?

This is to prevent people from submitting every other project of their own to `homebrew/core`. If your software is popular, it's fine.

**Reply by aryairani** [reply]
I wasn't sure from [Acceptable-Formulae](https://docs.brew.sh/Acceptable-Formulae) what was considered popular, but we have 3.8k stars, 185 forks, and 117 watchers, so I guess that means we've made it! 😁 🎉

**Reply by carlocab** [reply]
Our inclusion requirements are more modest than that. 🙂 Feel free to open a PR.

**Reply by aryairani** [reply]
@carlocab Related question: Is it acceptable / normal to open a PR and using it to debug the build process for an extended period? I recognize that a lot of the testing can be done up-front locally if you know all the right commands and have all the right hardware, but I'm 0 for 2. 😅   I assume I'll get better over time, though.  

Right now, I assume it won't pass CI due to various missing arm64 bottle dependencies, but someday I think it would be slick to get unison into the core repo.

**Reply by carlocab** [reply]
It's not normal. As for whether it's acceptable: maintainer opinions may vary. If it doesn't take too long to build (~less than 20 minutes for a full CI run) and you only push when there are no other CI jobs running, I'd say it's probably fine.

But ideally, you would resolve as many issues as possible locally (with `brew install -s your-formula && brew audit --new your-formula && brew test your-formula`) before trying to use Homebrew CI.

**Reply by SMillerDev** [reply]
> if you know all the right commands 

They're in the pull request template and in https://docs.brew.sh/Adding-Software-to-Homebrew

> and have all the right hardware

Homebrew also runs in Linux and wsl (we even have a docker container) so you can do a large part of the work there already. Should only be minor tweaks after.
